compiler/src/iree/compiler/Codegen/Common/WorkgroupSpecializationPass.cpp

```diff
@@ -0,0 +1,276 @@
+//   3. clone the body of the innermost loop in the else block.
+//   4. clone the body of the else block into the then block and update the
+//      bounded size ops with the constant tile size during the cloning.
+static void specializeDistributionLoops(
+    SmallVector<LoopTilingAndDistributionInfo> &infoVec) {
+  // get the scf for loop nests and their tile size.
+  SmallVector<scf::ForOp> distLoops;
+  SmallVector<int64_t> tileSizes;
+  SmallVector<AffineMinOp> minSizeOps;
+
+  // Distribution info vector has the innermost loop at index 0.
+  for (auto &info : infoVec) {
+    auto forOp = cast<scf::ForOp>(info.loop);
+    distLoops.push_back(forOp);
+    if (!info.tileSize) {
+      return;
+    }
+    int64_t tileSize = *info.tileSize;
+    tileSizes.push_back(tileSize);
+    SmallVector<AffineMinOp> boundedTileSizesOps =
+        getBoundedTileSizeOps(forOp, tileSize);
+    if (boundedTileSizesOps.empty()) {
+      // When there is no bounded tile size op, it means that the tensor
+      // size is already a multiple of tile size.
+      minSizeOps.push_back(AffineMinOp());
+    } else if (boundedTileSizesOps.size() != 1) {
+      // Supposed to see only one op for the bounded tile size if any.
+      return;
+    } else {
+      minSizeOps.push_back(boundedTileSizesOps[0]);
+    }
+  }
+
+  // Check the eligilbility. Unsupported cases are:
+  //   1. Dynamic cases
+  //   2. UB < Tile size
+  //   3. UB == a multiple of the tile size
+  bool isAlreadyMultiple = true;
+  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
+    scf::ForOp loop = distLoops[i];
+    auto ubOp = loop.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    if (!ubOp) {
+      // TODO: handle dynamic cases by adding more code to check the
+      // conditions.
+      return;
+    }
+    int64_t ub = ubOp.value();
+    int64_t tileSize = tileSizes[i];
+    if (ub % tileSize != 0) {
+      isAlreadyMultiple = false;
+    }
+    if (ub < tileSize) {
+      return;
+    }
+  }
+
+  if (isAlreadyMultiple) return;
+
+  scf::ForOp innermostLoop = distLoops[0];
+  auto loc = innermostLoop.getLoc();
+  Block *block = innermostLoop.getBody();
+
+  OpBuilder builder(innermostLoop->getContext());
+  PatternRewriter::InsertionGuard guard(builder);
+
+  AffineMinOp minOp0 = minSizeOps[0];
+  if (minOp0) {
+    // Make sure the affine.min in the innermost loop is the first
+    // instruction in the body.
+    auto minOp = llvm::dyn_cast_or_null<AffineMinOp>(block->front());
+    if (!minOp || minOp != minOp0) {
+      minOp0->moveBefore(&block->front());
+    }
+    builder.setInsertionPointAfter(&block->front());
+  } else {
+    builder.setInsertionPoint(&block->front());
+  }
+
+  // create a condition for scf.if
+  Value cond;
+  SmallVector<Value> constantOps;  // ConstantIndexOps for tile sizes
+  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
+    // clone the minSize op in the loop and place it before scf.if
```

Comment:
is this comment out of date? We don't clone anymore right?

Reply:
Fixed. Thanks